### ISSUE TITLE
Preferred idiom for exhaustiveness checking of unions

### ISSUE THREAD

**Issue by bluetech** [python/mypy#5818]
* Are you reporting a bug, or opening a feature request?

A question...

* Please insert below the code you are checking with mypy.

I have some union, e.g.

```python
from typing import Union

MyUnion = Union[int, str]
```

and I have a function which takes the union and matches on it:

```python
def handle_my_union(x: MyUnion) -> None:
    if isinstance(x, int): ...
    elif isinstance(x, str): ...
```

Now suppose that the union gets added a new member:

```diff
- MyUnion = Union[int, str]
+ MyUnion = Union[int, str, bytes]
```

I have to remember to add it to all of the places which match on the union, for example `handle_my_union`. As written above, mypy doesn't warn (and has no reason to). So what I use is this function (stolen mostly verbatim [from TypeScript](https://www.typescriptlang.org/docs/handbook/advanced-types.html#exhaustiveness-checking)):

```python
from typing import NoReturn

def assert_never(x: NoReturn) -> NoReturn:
    assert False, "Unhandled type: {}".format(type(x).__name__)
```

and use it as follows:

```diff
 def handle_my_union(x: MyUnion) -> None:
     if isinstance(x, int): ...
     elif isinstance(x, str): ...
+    else:
+        assert_never(x)
```

* What is the actual behavior/output?

Now mypy warns:

```
error: Argument 1 to "assert_never" has incompatible type "bytes"; expected "NoReturn"
```

When I add a branch to handle `bytes`:

```diff
 def handle_my_union(x: MyUnion) -> None:
     if isinstance(x, int): ...
     elif isinstance(x, str): ...
+    elif isinstance(x, bytes): ...
     else:
         assert_never(x)
```

the error is gone.

* What is the behavior/output you expect?

I am happy with this solution, I just wonder if this is something that is considered appropriate to rely on, or if there's another preferable way to achieve exhaustiveness checking?

* What are the versions of mypy and Python you are using?

Python 3.5.5, mypy 0.630.

* What are the mypy flags you are using? (For example --strict-optional)

`no_implicit_optional, check_untyped_defs, warn_unused_ignores`.

**Comment by srittau** [comment]
Typescript also has a `never` type that is not compatible to anything, not even itself. I wonder whether it makes sense to have something like this in `typing`, which could be used in place of the argument type here.

**Comment by bluetech** [comment]
Isn't `NoReturn` equivalent to `never`? I'm just concluding by analogy here, specifically

- In TypeScript, as you say, the `never` is used for the "no return" case, e.g. [node `process.exit`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/0e95a07e460e8b2223949a53fdcb5395b0bcda13/types/node/index.d.ts#L757).

- In Rust, the "no return" case actually came first (written as e.g. `pub fn exit(code: i32) -> !`) and was then generalized to a [never type](https://doc.rust-lang.org/nightly/std/primitive.never.html).

**Comment by gvanrossum** [comment]
This is a clever hack, and I would never have come up with that (probably because the use case is not one I commonly encounter). It makes sense to legitimize some form of this and document it.

**Comment by ethanhs** [comment]
I would also point out this would be inconsistent with PEP 484 which states that `NoReturn` can only be used as a return type:

>The NoReturn type is only valid as a return annotation of functions, and considered an error if it appears in other positions

Also, it might be nice to warn about using `NoReturn` in other places.

**Comment by bluetech** [comment]
> (Or I guess we could make NoReturn and Any interact in the same way they do in Typescript -- not sure how easy or hard that would be though.)

FWIW, after applying the following change, to which I reached with some printf debugging, mypy's test suite still passes, and the `NoReturn ← Any` assignment fails with "error: Incompatible types in assignment (expression has type "Any", variable has type "NoReturn")", like TypeScript does.

```diff
diff --git a/mypy/subtypes.py b/mypy/subtypes.py
index d06ad40f..c4a62a08 100644
--- a/mypy/subtypes.py
+++ b/mypy/subtypes.py
@@ -162,7 +162,7 @@ class SubtypeVisitor(TypeVisitor[bool]):
         return True
 
     def visit_any(self, left: AnyType) -> bool:
-        return True
+        return not isinstance(self.right, UninhabitedType)
 
     def visit_none_type(self, left: NoneTyp) -> bool:
         if experiments.STRICT_OPTIONAL:
```

**Comment by bluetech** [comment]
> I would also point out this would be inconsistent with PEP 484 which states that NoReturn can only be used as a return type:

The value of the argument of `assert_never` doesn't *really* matter as long as it doesn't match, I think. So `assert_never` can be defined like this instead:

```python
from typing import NoReturn
import enum

class Impossible(enum.Enum):
    pass

def assert_never(x: Impossible) -> NoReturn:
    assert False, "Unhandled type: {}".format(type(x).__name__)
```

**Comment by gvanrossum** [comment]
In that case you don't even have to inherit from Enum. You can just refrain
from creating instances. :-)

**Comment by ilevkivskyi** [comment]
> The NoReturn type is only valid as a return annotation of functions, and considered an error if it appears in other positions

I would actually propose to remove this restriction from the PEP (if it is still OK), and make `NoReturn` a normal type. Proposed pattern is sufficient motivation for this, I could imagine it is useful to statically assert that a block is unreachable, especially that mypy already can do this.

Regarding the subtyping between `Any` and `NoReturn` we have a separate issue https://github.com/python/mypy/issues/3194.

**Comment by ilevkivskyi** [comment]
The name is a bit unusual indeed, but typical alternatives `Uninhabited`, `BottomType`, etc. are not better IMO. Also after using it a bit, it actually kind of starts to make sense (at least for me).

**Comment by srittau** [comment]
Personally, I find Typescript's `never` type to be quite intuitive.

**Comment by bluetech** [comment]
I also like the name `Never`. I would go as far as deprecating `NoReturn` in favor of using `Never` for everything, like TypeScript does.

**Comment by JukkaL** [comment]
I my opinion `NoReturn` really only makes sense as a function return type. I think that `Never` could work better in other contexts. `Nothing` is another plausible name, but it can be confused with `None`.

I think that subtyping between `Any` and `Never` (or whatever it would be called) should be discussed separately from #3194, since that issue doesn't seem to be moving forward. Intuitively, `Any` is *some* value, whereas `NoReturn` is no value, so perhaps `Any` shouldn't be compatible with `NoReturn`.

**Comment by ilevkivskyi** [comment]
> I would go as far as deprecating `NoReturn` in favor of using `Never` for everything.

This is not going to happen, it's too late. There are around 30K files in GitHub that use `NoReturn`.

**Comment by ilevkivskyi** [comment]
> since that issue doesn't seem to be moving forward

Is this really an argument? I think it is always better to consider such edge cases in the context of the whole picture, otherwise it may be inconsistent. For example joins and meets should behave consistently w.r.t. to subtyping.

**Comment by JukkaL** [comment]
>  I think it is always better to consider such edge cases in the context of the whole picture, otherwise it may be inconsistent. 

Sure, but I'm not seeing how #3194 is directly relevant here.  #3194 proposes that we change the type system in certain ways, which may affect how `Any` and `NoReturn` interact. Of course any such proposed change needs to be internally consistent (and with other goals we have). However, it may be possible to change only how `Any`/`NoReturn` interact while otherwise keeping the current type system semantics, while becoming no less consistent. We don't know this yet, so I think that it's worth discussing. 

Or do you believe that the current type system is inconsistent, and any additional changes to the type system could make it harder to make it consistent in the future? If so, I'd like to see more examples of what is currently wrong.

**Comment by JukkaL** [comment]
> This is not going to happen, it's too late. 

We could perhaps rename `NoReturn` to `Never` and change `NoReturn` into an alias for `Never`? `NoReturn` could then be deprecated at some point. Not sure if this would be worth the possible confusion.

**Comment by ilevkivskyi** [comment]
> Not sure if this would be worth the possible confusion.

This is the main question. We can of course do this ("two step" deprecation), but the question is whether `Never` is really so much better. I agree this is a better name, bot not hugely better. So taking into account other priorities, I would not do this.

Another question to consider is how to repr this type. Currently it is `<nothing>`, which is probably OK, but if we are going forward should we change it to `<never>`? `List[<never>]` probably looks a bit weird, although we sometimes infer such types.

**Comment by bluetech** [comment]
[`assert_never`](https://docs.python.org/3.11/library/typing.html#typing.assert_never) and [`Never`](https://docs.python.org/3.11/library/typing.html#typing.Never) (an alias to `NoReturn`) will be [included](https://github.com/python/cpython/commit/243436f3779c1e7bed1fd4b23d5a8ec5eff40699) in Python 3.11 and are already included in [typing-extensions 4.1.0](https://mail.python.org/archives/list/typing-sig@python.org/thread/WIUT7GQQR7PEA6UTIWI7SDPGQJO2I5Z6/), thanks to @JelleZijlstra, so I believe this issue is settled!

**Comment by JelleZijlstra** [comment]
Correct! It's also documented in https://typing.readthedocs.io/en/latest/source/unreachable.html
